// awilix/dist/index.js
"use strict";
// These tests are not run, they are only here to verify typings.
Object.defineProperty(exports, "__esModule", { value: true });
//import awilix definitions
const index_1 = require("../index");
/**
 * Test class for container.
 * @class TestClass
 */
class TestClass {
    constructor(s, n) { }
    stuff(str) { }
}
/**
 * Test function for container.
 * @function testFunction
 */
function testFunction(str) { }
/**
 * Test value for container.
 */
const VALUE = 'foo';
const container = index_1.createContainer();
const scope = container.createScope();
container.register({
    testClass: index_1.asClass(TestClass),
    testClass2: index_1.asClass(TestClass),
    testFunction: index_1.asFunction(testFunction),
    testValue: index_1.asValue(VALUE)
});
container.cradle.testClass;
container.cradle.testFunction;
container.cradle.testValue;
container.registrations[0].resolve;
container.registrations[0].lifetime;
container.register('_testClass', index_1.asClass(TestClass));
const testClass = container.resolve('_testClass');
testClass.stuff("Hello");
const testFunc = container.resolve('_testFunction');
testFunc("");
const testFunc2 = container.resolve('_testFunction');
testFunc2("");
container.register('_testValue', index_1.asValue(VALUE));
container.registerClass({
    __testClass: TestClass
});
container.registerFunction({
    __testFunction: testFunction
});
container.registerValue({
    __testValue: VALUE
});
container.loadModules(['*.js'], {
    formatName: (name, descriptor) => descriptor.path
});
container.loadModules([
    ['hello.js', { lifetime: index_1.Lifetime.SCOPED }]
]);
index_1.listModules('');
index_1.listModules(['']);
index_1.listModules([
    ['hello.js', { lifetime: index_1.Lifetime.SCOPED }]
]);
